backend/src/resolvers.js

```diff
@@ -1,127 +1,315 @@
+      // if so, update value on vote
+      const doesPostExistAndIsUpvoted =
+        votes.length > 0 && votes[0].value == -1;
+
+      if (doesPostExistAndIsUpvoted) {
+        console.log('[upvote] post is downvotes, updating...');
+        // updat query
+        vote.where = { id: votes[0].id };
       }
 
-      return context.dataSources.db.upvotePost(post.id, context.req.auth.id);
+      console.log('[upvote] create a new vote for post');
+      // otherwise create a new vote
+
+      return await delegateToSchema({
+        schema,
+        operation: 'mutation',
+        fieldName: doesPostExistAndIsUpvoted ? 'updateVote' : 'createVote',
+        args: vote,
+        context,
+        info,
+      });
     },
 
-    downvote: (parent, args, context) => {
-      const post = context.dataSources.db.getPost(args.id);
+    downvote: async (_parent, args, context, info) => {
+      const {
+        data: { votes },
+      } = await executor({
+        document: QUERY_POST_VOTES,
+        variables: {
+          userId: context.person.id,
+          postId: args.post.id,
+        },
+      });
+
+      const vote = {
+        data: {
+          value: -1,
+          votedBy: {
+            connect: { id: context.person.id },
+          },
+          post: {
+            connect: { id: args.post.id },
+          },
+        },
+      };
+
+      // check if user has voted and post is downvoted (-1)#
+      // if so , do nothing
+      if (votes.length > 0 && votes[0].value == -1) {
+        console.log('[downvote] user has downvoted, doing nothing...');
+
+        return;
+      }
+
+      // check if user has voted and post is upvoted (1)
+      // if so, update value on vote
+      const doesPostExistAndIsUpvoted = votes.length > 0 && votes[0].value == 1;
+
+      if (doesPostExistAndIsUpvoted) {
+        console.log('[downvote] post is upvoted, updating...');
 
-      if (!post) {
-        throw new UserInputError('Invalid post', { invalidArgs: args.postId });
+        // updat query
+        vote.where = { id: votes[0].id };
       }
 
-      return context.dataSources.db.downvotePost(post.id, context.req.auth.id);
+      // otherwise create a new vote
+      return await delegateToSchema({
+        schema,
+        operation: 'mutation',
+        fieldName: doesPostExistAndIsUpvoted ? 'updateVote' : 'createVote',
+        args: vote,
+        context,
+        info,
+      });
     },
 
+    delete: async (_parent, args, context, info) => {
+      const {
+        data: { posts },
+      } = await executor({
+        document: QUERY_DOES_AUTHOR_HAS_POSTS,
+        variables: {
+          postId: args.post.id,
+          userId: context.person.id,
+        },
+      });
 
-    delete: (parent, args, context) => {
-      return context.dataSources.db.deletePost(args.id);
+      // throw error if no posts avilable
+      if (!posts.length) {
+        throw new UserInputError(
+          'This is not your post or post does not exist!'
```

Comment:
`This is not your post` should ideally be a permission in your `permissions.js` middleware. But I see that this might be problematic as it requires multiple GraphQL calls.